Upgrade to D3 v7

nglgzz opened the pull request:
This PR switches most of the visualization to use D3 v7 and to use the micro libraries instead of the entire D3 package. The only exception is the `queryPlan` component which is a bit trickier and would require quite a few changes since some of the functions it uses from D3 have been deprecated (I'll update that on another PR).

There are some minor differences in the updated visualization.

**Force simulation:**
- Precompute the force simulation instead of animating it on the first render. We do something like this now, but there still is some slight animation after rendering. This causes a noticeable performance improvement when rendering frames with lots of nodes + relationships.
- ~Added some centering forces to slightly decrease the scattering of nodes (making frames fit a few more nodes).~
- Pause simulation after 2 seconds of inactivity (the simulation is started on each interaction with the visualization). 
- The simulation is started when clicking a node. This is because the click triggers the `dragstart` event. I couldn't find a simple way to work around this, but I also didn't spend too much time looking into it. 

**Zoom:**
- Zoom buttons are not disabled when the min/max zoom is reached (I couldn't find a way to get this to work).
- I decreased the transition time for the zoom to make it feel snappier.
- Zoom can be controlled with shift + mouse wheel.

Below is a list of things I tested when comparing the current and the updated visualization.

**Force simulation:**

- The density and positioning of nodes, both when returning nodes + relationships and when returning only nodes.
- Distance at which nodes start repelling each other.

**Nodes/relationships:**

- Dragging nodes.
- Locking of nodes when dragging them.
- Selecting nodes/relationships shows information about them.
- Unselecting nodes/relationships by clicking on the "canvas".
- Using all options in the arc menu when clicking a node.
- Double click to expand/collapse a node.

**Frame:**

- Pan view
- Zoom buttons 
- Shift + scroll to zoom (only on the updated version)

nglgzz commented on src/browser/modules/D3Visualization/lib/visualization/renders/init.ts:
```diff
@@ -17,189 +17,166 @@
-    circles.attr({
-      r(node: VizNode) {
+      .data(node => [node])
+      .join('circle')
```
`.join` can now be used instead of `.enter().append()[...].exit().remove()`

FallingCeilingS commented on src/browser/modules/D3Visualization/lib/visualization/components/layout.ts:
```diff
@@ -59,52 +71,31 @@ const layout: AvailableLayouts = {
         return latestStats
       }
 
-      const accelerateLayout = function() {
```
@nglgzz - I'd vote for removing all unused collecting stats as it's not been analysed anywhere.

nglgzz replied:
I agree, I removed that code now.

FallingCeilingS reviewed (approved):
It looks good to me! 💯 It's much easier to [optimise the performance](https://medium.com/neo4j/scale-up-your-d3-graph-visualisation-webgl-canvas-with-pixi-js-63f119d96a28) on the top of your work. Great job!

OskarDamkjaer commented:
Nice to get get such an ancient version updated 😍 I've found two things to look into before merging; if I run `match (n) return n` with the movie database it seems the viz does not "finish" computing where the nodes should be, see video

https://user-images.githubusercontent.com/10564538/148796628-cba740ce-70a5-46e8-8784-09917d83e184.mp4

Also

It seems the "connect result nodes" setting is no longer respected (aka. theres no relatinoships added to MATCH (N) RETURN N).
![CleanShot 2022-01-10 at 16 54 46](https://user-images.githubusercontent.com/10564538/148797035-4c897ea4-c455-4d77-ba86-d5481910ccfb.png)
![CleanShot 2022-01-10 at 16 54 37](https://user-images.githubusercontent.com/10564538/148797042-24723d30-4cb6-4de8-a990-fe753456141a.png)

eijawerner reviewed (commented):
Great work! ⭐ So glad to see this upgrade finally getting in.

nglgzz commented:
@OskarDamkjaer I tweaked the initial position of nodes, so instead of arranging them in a circle, we now arrange them in a spiral. This should fix the issue you saw with `MATCH (n) RETURN n`. Below you can see the difference between the two initial layouts. Note that this is not how the nodes are shown when the frame is rendered, but their initial position before the simulation is started.

**Before**
![before](https://user-images.githubusercontent.com/13448636/148933040-d24ff523-d749-4f57-adb5-c276ca51d5cc.png)

**After**
![after](https://user-images.githubusercontent.com/13448636/148933049-cc8b91ec-65da-4d2d-906a-0f153eabc765.png)

nglgzz commented:
@FallingCeilingS 

> Note that this is not how the nodes are shown when the frame is rendered, but their initial position before the simulation is started.

This change is not visible, it's just the setup before the simulation is started. The screenshots are just to visualize what I'm talking about, you wouldn't actually see that.

nglgzz commented on src/browser/modules/D3Visualization/lib/visualization/components/Visualization.ts:
```diff
@@ -209,9 +214,9 @@ const vizFn = function (
-    .on('mousewheel.zoom', null as any)
+    .on('DOMMouseScroll.zoom', handleZoomOnShiftScroll)
+    .on('wheel.zoom', handleZoomOnShiftScroll)
+    .on('mousewheel.zoom', handleZoomOnShiftScroll)
```
I propose we merge as is and change later if needed. `shift+scroll` is not intrusive, so it shouldn't bother anyone if it's there and we change it later. Also, this way we'd get a build to test this implementation with no extra effort.

eijawerner replied:
Sounds good 👍 . Though it stills competes with the full page scroll when the shift key is pressed so please add a e.preventDefault() in the handleZoomOnShiftScroll before merging.

nglgzz replied:
Good point, the e.preventDefault is now there :sparkles:

nglgzz commented:
> @nglgzz - regarding `MATCH (n) RETURN n` problem, does it relate to you set the simulation timeout?

Sort of. I think there are three parts to this issue:
1. The initial layout places nodes quite far apart, so naturally it takes a while before they all get pushed towards the center.
2. I removed the accelerateLayout function, which seemed to be doing something similar to the precomputing I added, but it also was overriding the tick function from d3 to do multiple ticks at once (based on this and the name, I guess the idea here is to make the simulation run faster, but there are better ways to do this now). 
3. I added a timeout that will stop the simulation after two seconds, so if it's taking too long for the nodes to get pushed to the center, they'll just stop outside of the view. 

By changing the initial layout to pack the nodes more densely and to start placing them in the center, points 2 and 3 should effectively no longer be an issue.

OskarDamkjaer commented:
I've tested the newest changes and it seems like the spiral set-up has unintended consequences on the node layout. match (n) return n on the movie database now shows this:
![CleanShot 2022-01-11 at 13 30 27](https://user-images.githubusercontent.com/10564538/148943409-d0838f69-6dab-4146-be40-64c80f04a47b.png)
instead of this: 
![CleanShot 2022-01-11 at 13 30 32](https://user-images.githubusercontent.com/10564538/148943440-413b5d61-b276-4703-9c16-0fc09862149a.png)

If I click a node to let it compute calcuate the force for a little while longer I get this: 
![CleanShot 2022-01-11 at 13 33 42](https://user-images.githubusercontent.com/10564538/148943560-8bddb318-4516-4382-8519-17912169c151.png)

Both of which are more entangled/worse

nglgzz commented:
Nice catch, I didn't notice that. I'm looking into it and so far it seems like it should be just a matter of tweaking the forces, in particular the `forceManyBody` (which should be equivalent to `charge` in the previous version). When I try to set the same value we had before (-1000 compared to the default value which is -30) the simulation becomes super slow and I start getting a ton of errors in the console, but it does look closer to the old visualization. Will play around a bit more to see if I can get around this.

nglgzz commented:
I reverted the changes to the initial layout and tweaked `forceManyBody` and the centering forces. Now the positioning should be closer to the original visualization.

I also realized instead of 2 seconds I set 1.5 for the timeout of the simulation  (I was testing that change and forgot to revert it :smile:), so I changed that to actually be 2 seconds.

eijawerner commented:
Tested the latest out and it seems that the repelling forces between unconnected nodes are too large so they keep move out of sight for ever in this case:

https://user-images.githubusercontent.com/11065688/149087785-5a5024cd-63de-44b4-9ed6-25613f24258a.mov

eijawerner commented on src/browser/modules/D3Visualization/lib/visualization/utils/circularLayout.ts:
```diff
@@ -19,7 +19,7 @@
 import VizNode from '../components/VizNode'
 
-export default function circularLayout(
+export default function spiralLayout(
```
You forgot to change the name back.

eijawerner commented on src/browser/modules/D3Visualization/lib/visualization/components/layout.ts:
```diff
@@ -115,15 +91,43 @@ const layout: AvailableLayouts = {
+          )
+          .force('center', forceCenter(center.x, center.y))
+          // Centering forces for nodes with no relationships to prevent them
+          // from getting pushed out of view by the forceManyBody force.
```
Change this comment or remove it now when all nodes have the same centering forces.

FallingCeilingS commented:
I noticed that when rendering large amount of nodes, probably due to simulation timeout, nodes became more overlapping (because we don't allow them complete the simulation).

Before (the simulation is slow but nodes are mostly not overlapped at the end):
<img width="1186" alt="Screenshot 2022-01-12 at 14 14 58" src="https://user-images.githubusercontent.com/26452483/149156765-16f4f84c-8162-41fb-8903-8bcf6ada6b90.png">

After (the simulation stopped within 2 seconds but nodes are more overlapped):
<img width="1696" alt="Screenshot 2022-01-12 at 14 16 05" src="https://user-images.githubusercontent.com/26452483/149156914-ef8e8a96-b927-4439-a608-930a0f5b4d52.png">


Is there any way to reduce overlapping by configuring `force`?

eijawerner reviewed (approved):
I have tested out the latest and it looks good now as far as the examples I have tested. Good work! I have some really minor comments that you could fix if you want.

FallingCeilingS commented:
The nodes are more overlapped compared to the previous version when rendering large number of them (i.e. 1000 nodes):

Before: the simulation is slow, but the overlapping is less:
<img width="1186" alt="Screenshot 2022-01-12 at 14 14 58" src="https://user-images.githubusercontent.com/26452483/149158774-5aa01670-7685-4082-921a-402fe5477257.png">

After: the simulation is quick, but it seems like it doesn't have enough time to avoid overlapping:
<img width="1693" alt="Screenshot 2022-01-12 at 14 28 02" src="https://user-images.githubusercontent.com/26452483/149158973-28db3146-f948-416e-a52a-5c4e29383331.png">

Is it possible to config `force` simulation to reduce overlapping? If not, we can just leave it for now and see if drawing them on canvas helps us use more time to simulate with no performance issues.

Also, I've played around with interactions, like dragging, zooming, double-clicking, etc. The performance is generally the same as before, but setting simulation timeout indeed makes the app more responsive!  💯

nglgzz commented:
@FallingCeilingS There is already a collision force, either increasing its strength or running the simulation for longer should make a difference. 

However, I think with this number of nodes/relationships the visualization isn't actually useful (as it's too crowded in both cases) and the difference is not that noticeable, so I'd prefer having the app be more responsive over a slightly less dense visualization. I agree that this can probably be addressed once we switch to canvas though.

OskarDamkjaer commented:
I've not given the latest version a spin but the code looks good and I trust Eija&Juns testing for the force layout, anything left to do before merging?
